### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by meyerar**
This pull request contains the following:
 - Some minior code refactoring
 - XML Comments on some of the inportant objects
 - relocated the Document reference builder to FhirFactory
for FHIR:
 - Removed the fullUrl on the resources (was not needed)
 - System attribute is set as "_" (we do not have that those types defined yet)
 - aligned the FHIR profile to Hiso (see attached document)
 
This functionality has been deployed to sandbox account (DEV-MEYERA-X & INT-X-STD)
`
{
    "resourceType": "Bundle",
    "id": "2a342db9-f15f-4feb-8318-a3dc0e50f8de",
    "meta": {
        "lastUpdated": "2023-05-11T04:36:22.7234705+00:00"
    },
    "type": "searchset",
    "link": [
        {
            "relation": "self",
            "url": "https://preprod-apibroker.whanautahi.com/mdr/fhir/MedicationRequest?patient=MTA7448"
        }
    ],
    "entry": [
        {
            "fullUrl": "https://preprod-apibroker.whanautahi.com/mdr/fhir/MedicationRequest/2CP1T1BHK2V322J764-1",
            "resource": {
                "resourceType": "MedicationRequest",
                "id": "2CP1T1BHK2V322J764-1",
                "meta": {
                    "versionId": "1",
                    "lastUpdated": "2021-07-21T21:06:03.413+00:00",
                    "source": "c3677efe-2234-4bac-883c-4d0ca9abd681",
                    "security": [
                        {
                            "system": "http://terminology.hl7.org/CodeSystem/v3-Confidentiality",
                            "code": "U",
                            "display": "unrestricted"
                        }
                    ]
                },
                "extension": [
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-status-type",
                        "valueCode": "Valid"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-action",
                        "valueCode": "None"
                    },
                    {
                        "extension": [
                            {
                                "url": "endorsementType",
                                "valueCodeableConcept": {
                                    "coding": [
                                        {
                                            "system": "http://hl7.org.nz/fhir/CodeSystem/nzeps-endorsement-type",
                                            "code": "CERTCOND",
                                            "display": "Certified Condition"
                                        }
                                    ]
                                }
                            },
                            {
                                "url": "endorsementValue",
                                "valueString": "yes"
                            },
                            {
                                "url": "expiryDate",
                                "valueDateTime": "2022-02-07T08:59:00+13:00"
                            }
                        ],
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-endorsement"
                    },
                    {
                        "extension": [
                            {
                                "url": "endorsementType",
                                "valueCodeableConcept": {
                                    "coding": [
                                        {
                                            "system": "http://hl7.org.nz/fhir/CodeSystem/nzeps-endorsement-type",
                                            "code": "FD",
                                            "display": "Frequent Dispensing"
                                        }
                                    ]
                                }
                            },
                            {
                                "url": "endorsementValue",
                                "valueString": "every 120 days"
                            },
                            {
                                "url": "expiryDate",
                                "valueDateTime": "2021-11-19T08:59:00+13:00"
                            }
                        ],
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-endorsement"
                    },
                    {
                        "extension": [
                            {
                                "url": "endorsementType",
                                "valueCodeableConcept": {
                                    "coding": [
                                        {
                                            "system": "http://hl7.org.nz/fhir/CodeSystem/nzeps-endorsement-type",
                                            "code": "TRPER",
                                            "display": "Trial Period"
                                        }
                                    ]
                                }
                            },
                            {
                                "url": "endorsementValue",
                                "valueString": "15 days"
                            },
                            {
                                "url": "expiryDate",
                                "valueDateTime": "2021-08-06T08:59:00+12:00"
                            }
                        ],
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-endorsement"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-instructions-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-long-term-medication",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-unusual-dose-quantity",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-supply-period-reason",
                        "valueString": "Patient going overseas for 3 months"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-trial-period-reason",
                        "valueString": "To confirm no adverse reaction"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-funding-category",
                        "valueString": "BSO"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-claimable",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-private-item",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-force-notification",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-notify",
                        "valueBoolean": false
                    }
                ],
                "identifier": [
                    {
                        "system": "https://standards.digital.health.nz/ns/nzeps-scid-item-id",
                        "value": "2CP1T1BHK2V322J764-1"
                    },
                    {
                        "system": "https://www.gs1.org/gtin",
                        "value": "912312312380109280"
                    }
                ],
                "status": "active",
                "intent": "order",
                "medicationCodeableConcept": {
                    "coding": [
                        {
                            "system": "http://nzmt.org.nz",
                            "code": "31108311000116105",
                            "display": "paliperidone 25 mg/0.25 mL injection: modified release, 1 x 0.25 mL syringe"
                        }
                    ],
                    "text": "Olive Oil 10% in Aqueous Cream"
                },
                "subject": {
                    "reference": "Patient/MTA7448",
                    "type": "Patient",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/nhi-id",
                        "value": "MTA7448"
                    },
                    "display": "Mr Bill Gates | Male | 1966-07-22"
                },
                "supportingInformation": [
                    {
                        "reference": "Location/FB0326-G",
                        "type": "Location",
                        "identifier": {
                            "system": "https://standards.digital.health.nz/ns/hpi-facility-id",
                            "value": "FB0326-G"
                        },
                        "display": "Jay Leno's Garage"
                    }
                ],
                "authoredOn": "2021-07-22T08:59:00+12:00",
                "requester": {
                    "reference": "Practitioner/FYF886-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FYF886-C"
                    },
                    "display": "Prescriber Leno | Doctor | 99109"
                },
                "recorder": {
                    "reference": "Practitioner/FMG861-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FMG861-C"
                    },
                    "display": "Mrs Caitlyn Jenner MD"
                },
                "note": [
                    {
                        "text": "Prescribers Comments: reduce inflammation"
                    },
                    {
                        "text": "Reason for Prescribing: back spasms"
                    }
                ],
                "dosageInstruction": [
                    {
                        "sequence": 1,
                        "patientInstruction": "Take 1 tablet, 4 times a day",
                        "timing": {
                            "repeat": {
                                "frequency": 6,
                                "period": 1,
                                "periodUnit": "d"
                            }
                        },
                        "route": {
                            "coding": [
                                {
                                    "system": "http://terminology.hl7.org/CodeSystem/v3-RouteOfAdministration",
                                    "code": "PO",
                                    "display": "Swallow, oral"
                                }
                            ]
                        },
                        "doseAndRate": [
                            {
                                "type": {
                                    "coding": [
                                        {
                                            "system": "http://terminology.hl7.org/CodeSystem/dose-rate-type",
                                            "code": "ordered",
                                            "display": "Ordered"
                                        }
                                    ]
                                },
                                "doseQuantity": {
                                    "value": 99,
                                    "unit": "Tablet",
                                    "code": "428673006"
                                }
                            }
                        ]
                    }
                ],
                "dispenseRequest": {
                    "dispenseInterval": {
                        "value": 0,
                        "unit": "d"
                    },
                    "validityPeriod": {
                        "start": "2021-07-22",
                        "end": "2021-11-22"
                    },
                    "numberOfRepeatsAllowed": 3,
                    "quantity": {
                        "value": 100,
                        "unit": "mg"
                    }
                },
                "substitution": {
                    "allowedBoolean": true
                }
            },
            "search": {
                "mode": "match"
            }
        },
        {
            "fullUrl": "https://preprod-apibroker.whanautahi.com/mdr/fhir/MedicationRequest/2CP1T1BHK2V322J764-2",
            "resource": {
                "resourceType": "MedicationRequest",
                "id": "2CP1T1BHK2V322J764-2",
                "meta": {
                    "versionId": "1",
                    "lastUpdated": "2021-07-21T21:06:03.413+00:00",
                    "source": "c3677efe-2234-4bac-883c-4d0ca9abd681",
                    "security": [
                        {
                            "system": "http://terminology.hl7.org/CodeSystem/v3-Confidentiality",
                            "code": "U",
                            "display": "unrestricted"
                        }
                    ]
                },
                "extension": [
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-status-type",
                        "valueCode": "Valid"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-action",
                        "valueCode": "None"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-instructions-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-long-term-medication",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-unusual-dose-quantity",
                        "valueBoolean": true
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-funding-category",
                        "valueString": "BSO"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-claimable",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-private-item",
                        "valueBoolean": true
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-force-notification",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-notify",
                        "valueBoolean": false
                    }
                ],
                "identifier": [
                    {
                        "system": "https://standards.digital.health.nz/ns/nzeps-scid-item-id",
                        "value": "2CP1T1BHK2V322J764-2"
                    },
                    {
                        "system": "https://www.gs1.org/gtin",
                        "value": "953213312332131231"
                    }
                ],
                "status": "active",
                "intent": "order",
                "medicationCodeableConcept": {
                    "coding": [
                        {
                            "system": "http://nzmt.org.nz",
                            "code": "20016811000116103",
                            "display": "risperidone 25 mg injection: modified release [1 x 25 mg vial] (&) inert substance diluent [1 x 2 mL syringe], 1 pack"
                        }
                    ],
                    "text": "Apple 10% in Orange"
                },
                "subject": {
                    "reference": "Patient/MTA7448",
                    "type": "Patient",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/nhi-id",
                        "value": "MTA7448"
                    },
                    "display": "Mr Bill Gates | Male | 1966-07-22"
                },
                "supportingInformation": [
                    {
                        "reference": "Location/FB0326-G",
                        "type": "Location",
                        "identifier": {
                            "system": "https://standards.digital.health.nz/ns/hpi-facility-id",
                            "value": "FB0326-G"
                        },
                        "display": "Jay Leno's Garage"
                    }
                ],
                "authoredOn": "2021-07-22T08:59:00+12:00",
                "requester": {
                    "reference": "Practitioner/FYF886-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FYF886-C"
                    },
                    "display": "Prescriber Leno | Doctor | 99109"
                },
                "recorder": {
                    "reference": "Practitioner/FMG861-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FMG861-C"
                    },
                    "display": "Mrs Caitlyn Jenner MD"
                },
                "note": [
                    {
                        "text": "Prescribers Comments: reduce inflammation"
                    },
                    {
                        "text": "Reason for Prescribing: back spasms"
                    }
                ],
                "dosageInstruction": [
                    {
                        "sequence": 1,
                        "patientInstruction": "Take 5ml, 4 times a day",
                        "timing": {
                            "repeat": {
                                "frequency": 4,
                                "period": 1,
                                "periodUnit": "d"
                            }
                        },
                        "route": {
                            "coding": [
                                {
                                    "system": "http://terminology.hl7.org/CodeSystem/v3-RouteOfAdministration",
                                    "code": "PO",
                                    "display": "Swallow, oral"
                                }
                            ]
                        },
                        "doseAndRate": [
                            {
                                "type": {
                                    "coding": [
                                        {
                                            "system": "http://terminology.hl7.org/CodeSystem/dose-rate-type",
                                            "code": "ordered",
                                            "display": "Ordered"
                                        }
                                    ]
                                },
                                "doseQuantity": {
                                    "value": 99,
                                    "code": "428673006"
                                }
                            }
                        ]
                    }
                ],
                "dispenseRequest": {
                    "dispenseInterval": {
                        "value": 0,
                        "unit": "d"
                    },
                    "numberOfRepeatsAllowed": 3,
                    "quantity": {
                        "value": 200,
                        "unit": "ml"
                    }
                },
                "substitution": {
                    "allowedBoolean": true
                }
            },
            "search": {
                "mode": "match"
            }
        },
        {
            "fullUrl": "https://preprod-apibroker.whanautahi.com/mdr/fhir/MedicationRequest/2CP1T1BHK2V322J764-3",
            "resource": {
                "resourceType": "MedicationRequest",
                "id": "2CP1T1BHK2V322J764-3",
                "meta": {
                    "versionId": "1",
                    "lastUpdated": "2021-07-21T21:06:03.413+00:00",
                    "source": "c3677efe-2234-4bac-883c-4d0ca9abd681",
                    "security": [
                        {
                            "system": "http://terminology.hl7.org/CodeSystem/v3-Confidentiality",
                            "code": "U",
                            "display": "unrestricted"
                        }
                    ]
                },
                "extension": [
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-status-type",
                        "valueCode": "Valid"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-action",
                        "valueCode": "None"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-instructions-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-long-term-medication",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-modified",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-unusual-dose-quantity",
                        "valueBoolean": true
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-funding-category",
                        "valueString": "BSO"
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-medication-claimable",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-private-item",
                        "valueBoolean": true
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-force-notification",
                        "valueBoolean": false
                    },
                    {
                        "url": "http://hl7.org.nz/fhir/StructureDefinition/nzeps-notify",
                        "valueBoolean": false
                    }
                ],
                "identifier": [
                    {
                        "system": "https://standards.digital.health.nz/ns/nzeps-scid-item-id",
                        "value": "2CP1T1BHK2V322J764-3"
                    },
                    {
                        "system": "https://www.gs1.org/gtin",
                        "value": "953213312332131231"
                    }
                ],
                "status": "active",
                "intent": "order",
                "medicationCodeableConcept": {
                    "coding": [
                        {
                            "system": "http://nzmt.org.nz",
                            "code": "20061161000116100",
                            "display": "olanzapine (as embonate monohydrate) 210 mg injection: modified release [1 x 210 mg vial] (&) inert substance diluent [1 x 3 mL vial]"
                        }
                    ],
                    "text": "Apple 10% in Orange"
                },
                "subject": {
                    "reference": "Patient/MTA7448",
                    "type": "Patient",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/nhi-id",
                        "value": "MTA7448"
                    },
                    "display": "Mr Bill Gates | Male | 1966-07-22"
                },
                "supportingInformation": [
                    {
                        "reference": "Location/FB0326-G",
                        "type": "Location",
                        "identifier": {
                            "system": "https://standards.digital.health.nz/ns/hpi-facility-id",
                            "value": "FB0326-G"
                        },
                        "display": "Jay Leno's Garage"
                    }
                ],
                "authoredOn": "2021-07-22T08:59:00+12:00",
                "requester": {
                    "reference": "Practitioner/FYF886-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FYF886-C"
                    },
                    "display": "Prescriber Leno | Doctor | 99109"
                },
                "recorder": {
                    "reference": "Practitioner/FMG861-C",
                    "type": "Practitioner",
                    "identifier": {
                        "system": "https://standards.digital.health.nz/ns/hpi-person-id",
                        "value": "FMG861-C"
                    },
                    "display": "Mrs Caitlyn Jenner MD"
                },
                "note": [
                    {
                        "text": "Prescribers Comments: reduce inflammation"
                    },
                    {
                        "text": "Reason for Prescribing: back spasms"
                    }
                ],
                "dosageInstruction": [
                    {
                        "sequence": 1,
                        "patientInstruction": "Take 5ml, 4 times a day",
                        "timing": {
                            "repeat": {
                                "frequency": 4,
                                "period": 1,
                                "periodUnit": "d"
                            }
                        },
                        "route": {
                            "coding": [
                                {
                                    "system": "http://terminology.hl7.org/CodeSystem/v3-RouteOfAdministration",
                                    "code": "PO",
                                    "display": "Swallow, oral"
                                }
                            ]
                        },
                        "doseAndRate": [
                            {
                                "type": {
                                    "coding": [
                                        {
                                            "system": "http://terminology.hl7.org/CodeSystem/dose-rate-type",
                                            "code": "ordered",
                                            "display": "Ordered"
                                        }
                                    ]
                                },
                                "doseQuantity": {
                                    "value": 99,
                                    "code": "428673006"
                                }
                            }
                        ]
                    }
                ],
                "dispenseRequest": {
                    "dispenseInterval": {
                        "value": 0,
                        "unit": "d"
                    },
                    "numberOfRepeatsAllowed": 3,
                    "quantity": {
                        "value": 200,
                        "unit": "ml"
                    }
                },
                "substitution": {
                    "allowedBoolean": true
                }
            },
            "search": {
                "mode": "match"
            }
        }
    ]
}`

### CODE REVIEW

File: src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs [hunk 1]
```diff
@@ -1,20 +1,67 @@
...
 public class AWSCloudWatchConfiguration
 {
+    /// <summary>
+    /// This is the Sectiom defined in aws parameter store as CloudWatchConfiguration/config
+    /// </summary>
     public const string SectionName = "CloudWatchConfiguration";
+    /// <summary>
+    /// config is there to pull stored json from parameter store / appsettings
+    /// </summary>
     public string config { get; set; }
```
Comment (wliuhealthnz) [inline src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs:18]: Can we make it 'Config' to be consistent?

File: src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs [hunk 2]
```diff
@@ -1,20 +1,67 @@
...
+    /// </summary>
     public bool CreateLogGroup { get; set; }
+    /// <summary>
+    /// optional parameter Retry Attempts
+    /// </summary>
     public byte RetryAttempts { get; set; }
+    /// <summary>
+    /// This parameter defines the System where the logger is implemented
+    /// </summary>
     public SystemType? type { get; set; }
```
Comment (wliuhealthnz) [inline src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs:58]: 'Type' to be consistent.

File: src/HealthShare.Referrals.Application/Commands/Services/EventStoreMapper.cs [hunk 3]
```diff
@@ -7,15 +7,26 @@
...
+    /// <summary>
+    /// EventStoreMapper
+    /// </summary>
+    /// <param name="loggerService"> LoggerService</param>
+    /// <param name="fhirFactory">factory to provide mapping for fhir projection</param>
+    /// <param name="datastoreFactory">factory to provide mapping for DataStore attrivuted</param>
+    /// <param name="jSONParser"></param>
     public EventStoreMapper(ILoggerService loggerService,
         IFhirFactory fhirFactory
         ,IDataStoreFactory datastoreFactory,
```
Comment (wliuhealthnz) [inline src/HealthShare.Referrals.Application/Commands/Services/EventStoreMapper.cs:32]: comma at the end will be consistent.

File: src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs (hunk 1 above)
Comment (meyerar) [reply src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs]: This has been implemented in the latest push

File: src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs (hunk 2 above)
Comment (meyerar) [reply src/HealthShare.Common.Core.Shared/Configurations/AWSCloudWatchConfiguration.cs]: This was implemented in the last push

**Review by wliuhealthnz — APPROVED** [review]

**Pull request merged**